modules/ipp/scanner.go

```diff
@@ -263,99 +218,186 @@ type Attribute struct {
 	ValueTag byte  `json:"tag,omitempty"`
 }
 
-type AttributeGroup struct {
-	Attrs []*Attribute
-	Tag byte
+func shouldReturnAttrs(length, soFar, size, upperBound int) (bool, error) {
+	if soFar + length > size {
+		if size == upperBound {
+			return true, zgrab2.NewScanError(zgrab2.SCAN_PROTOCOL_ERROR, errors.New("Reported field length runs out of bounds."))
+		}
+		return true, nil
+	}
+	return false, nil
 }
 
-func printGroup(g *AttributeGroup) {
-	fmt.Printf("0x%x\n", g.Tag)
-	for _, attrP := range g.Attrs {
-		fmt.Printf("\t0x%x: %q\n", attrP.ValueTag, attrP.Name)
-		for _, v := range attrP.Values {
-			fmt.Printf("\t\t%v\n", v)
-		}
+func detectReadBodyError(err error) error {
+	if err == io.EOF || err == io.ErrUnexpectedEOF {
+		return zgrab2.NewScanError(zgrab2.SCAN_PROTOCOL_ERROR, errors.New("Fewer body bytes read than expected."))
 	}
+	return zgrab2.NewScanError(zgrab2.TryGetScanStatus(err), err)
 }
 
-// TODO: Log every error that could come out of this
-// TODO: Determine whether errors should be ignored, debug logged, fatal, etc.
-func readAllAttributes(body *[]byte) (*[]*Attribute) {
-	buf := bytes.NewBuffer(*body)
+/* An IPP response contains the following data (as specified in RFC 8010 Section 3.1.8
+   https://tools.ietf.org/html/rfc8010#section-3.1.8)
+bytes name
+----------------------------
+2     version-number
+2     status-code
+4     request-id
+
+(0 or more instances of the following pair of fields)
+1     delimiter-tag OR value-tag
```

Comment:
This is really helpful. I feel like the only other information I would need to eyeball a response body is how to distinguish a value-tag from a delimiter-tag.